docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+
+This approach uses the `Node`'s `status.conditions` as the standard way to publish and consume different types (both [standard](https://github.com/kubernetes/node-problem-detector/blob/master/vendor/k8s.io/api/core/v1/types.go#L4157) and [custom](https://github.com/kubernetes/node-problem-detector/tree/master/config)) of health conditions.
+
+It also separates the components that [monitor](https://github.com/kubernetes/node-problem-detector/tree/master/config) the health conditions from those that react and take the [remedial action](https://github.com/kubernetes/node-problem-detector#remedy-systems).
+
+### Cluster Registry
+
+The [cluster-registry](https://github.com/kubernetes/cluster-registry) is intended as a light-weight way to keep track of a list of Kubernetes clusters along with some metadata.
+
+It defines the [`Cluster`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L30) [`CustomResourceDefinition`](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) for that purpose. The resource definition includes a [`status.conditions`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L66) field which can contain an arbitrary number (and [type](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L133)) of [`ClusterCondition`s](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L147).
+
+## Proposal
+
+Extending the analogy of [node-problem-detector](#node-problem-detector) to the larger granularity of the Kubernetes cluster, the [`ClusterCondition`s](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L147) in the[`status`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L66) of the [`Cluster`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L30) custom resource (as defined in the [cluster-registry](#cluster-registry)) can be used as the standard way to publish and consume health conditions that have cluster-wide implication.
+
+I.e.
+- A single `Cluster` resource instance is maintained for each cluster in the shoot control-plane namespace of the seed cluster.
+- Many different monitoring components (either dedicated to each shoot cluster or deployed as a common component in the seed clusters) then update the `status.conditions` on the `Cluster` resource instance(s).
+- Many different remedy systems watch the `status.conditions` (either dedicated to each shoot cluster or deployed as a common component in the seed cluster) on the `Cluster` resource instance(s) for conditions of (one or more) specific condition types and take remedial action(s). 
```

Comment:
> This has to be aligned with the Gardener extensibility, cause not all shoots have the same control plane,

Of course.

>  so these remedy controllers must be as much as possible cloud agnostic.

For most of the examples I have mentioned, the actions are cloud agnostic. However, I am not ruling out cloud provider aware conditions and actions in the future (as mentioned [above](https://github.com/gardener/gardener/pull/1797/files#r367814260)). Those must be aligned with Gardener extensibility, of course.

But this proposal is about the approach rather than the specific actions. Though the conditions and actions of [immediate relevance](https://github.com/gardener/gardener/pull/1797/files#diff-dc2b4cab766fb3ad06a4e681a236b570R104) could be picked as part of the first implementation. This is because these actions are already implemented and live.